### PR TITLE
VM:Implement THROW opcode

### LINKED ISSUE #193: VM: Implement Throw Opcode

### PR DESCRIPTION
**Pull request by dauTT**
Closes #193

### CODE REVIEW

**Comment by dauTT** [comment]
@decentralisedkev , It seems to me that Travis is not working correctly. The tests should all pass.

File: pkg/vm/vm_test.go [hunk 1]
```diff
@@ -91,6 +91,61 @@ func TestSimpleRun(t *testing.T) {
...
+
+	builder := stack.NewBuilder()
+	builder.EmitInt(20).EmitOpcode(stack.THROW)
+
+	// Pass program to VM
+	vm := NewVM(builder.Bytes())
+
+	// Runs vm with program
+	_, err := vm.Run()
+	assert.Equal(t, "the execution of the script program end with an error", err.Error())
```
Comment (decentralisedkev) [inline pkg/vm/vm_test.go]: You should not compare errors using their strings; it's very fragile. For now, just check for an error. In a future PR, a custom error struct can implemented for the VM.

File: pkg/vm/vm_test.go [hunk 2]
```diff
@@ -91,6 +91,61 @@ func TestSimpleRun(t *testing.T) {
...
+
+	ctx, err := vm.InvocationStack.CurrentContext()
+	assert.Equal(t, nil, err)
+	assert.Equal(t, 1, ctx.Estack.Len())
+	assert.Equal(t, -1, ctx.Astack.Len())
+}
+
+func TestThrowIfNot(t *testing.T) {
```
Comment (decentralisedkev) [inline pkg/vm/vm_test.go]: This test is the same as TestSimpleRun. Let's just rename TestSimpleRun to TestThrowIfNot

File: pkg/vm/vm_test.go [hunk 3]
```diff
@@ -91,6 +91,61 @@ func TestSimpleRun(t *testing.T) {
...
+	builder := stack.NewBuilder()
+	builder.EmitInt(20).EmitInt(20).EmitOpcode(stack.ADD)
+	builder.EmitInt(54).EmitOpcode(stack.EQUAL).EmitOpcode(stack.THROWIFNOT)
+
+	// Pass program to VM
+	vm := NewVM(builder.Bytes())
+
+	// Runs vm with program
+	_, err := vm.Run()
+	assert.Equal(t, "item on top of stack evaluates to false", err.Error())
```
Comment (decentralisedkev) [inline pkg/vm/vm_test.go]: same as above about comparing error strings

File: pkg/vm/vm_test.go (hunk 1 above)
Comment (dauTT) [reply pkg/vm/vm_test.go]: :+1:

File: pkg/vm/vm_test.go (hunk 2 above)
Comment (dauTT) [reply pkg/vm/vm_test.go]: It is not really the same. 

TestSimpleRun:
20 + 34 = 54

TestThrowIfNot:
20 + 20 = 54

File: pkg/vm/vm_test.go (hunk 3 above)
Comment (dauTT) [reply pkg/vm/vm_test.go]: :+1:

File: pkg/vm/vm_test.go (hunk 2 above)
Comment (decentralisedkev) [reply pkg/vm/vm_test.go]: Didn't see the change in number. They are both testing the THROWIFNOT function, one should throw and other should not. I don't think we need two separate test functions for this

**Comment by decentralisedkev** [comment]
LGTM

**Comment by hal0x2328** [comment]
50 points dauTT

**Comment by lllwvlvwlll** [comment]
https://neoscan.io/transaction/630c74684cc6dad330397c63e8c7aa6403c043cc1f241638cff87e3b97a40716

**Pull request merged**